src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -50,153 +51,137 @@ case class InferReadWriteAnnotation(t: String, tID: TransID)
-        args(1).serialize == b.serialize && we(args(0)) == zero
-      case _ => false
+  def replaceExp(repl: Netlist)(e: Expression): Expression =
+    e map replaceExp(repl) match {
+      case e: WSubField => repl getOrElse (e.serialize, e)
+      case e => e
     }
 
-    def inferReadWrite(s: Statement): Statement = s map inferReadWrite match {
-      // infer readwrite ports only for non combinational memories
-      case mem: DefMemory if mem.readLatency > 0 =>
-        val bt = UIntType(IntWidth(1))
-        val ut = UnknownType
-        val ug = UNKNOWNGENDER
-        val readers = HashSet[String]()
-        val writers = HashSet[String]()
-        val readwriters = ArrayBuffer[String]()
-        for (w <- mem.writers ; r <- mem.readers) {
-          val wp = getProductTerms(s"${mem.name}.$w.en")
-          val rp = getProductTerms(s"${mem.name}.$r.en")
-          if (wp exists (a => rp exists (b => checkComplement(a, b)))) {
-            val allPorts = (mem.readers ++ mem.writers ++ mem.readwriters ++ readwriters).toSet
-            // Uniquify names by examining all ports of the memory
-            var rw = (for {
-                idx <- Stream from 0
-                newName = s"rw_$idx"
-                if !allPorts(newName)
-              } yield newName).head
-            val rw_exp = WSubField(WRef(mem.name, ut, MemKind, ug), rw, ut, ug)
-            readwriters += rw
-            readers += r
-            writers += w
-            repl(s"${mem.name}.$r.en")   = EmptyExpression
-            repl(s"${mem.name}.$r.clk")  = EmptyExpression
-            repl(s"${mem.name}.$r.addr") = EmptyExpression
-            repl(s"${mem.name}.$r.data") = WSubField(rw_exp, "rdata", mem.dataType, MALE)
-            repl(s"${mem.name}.$w.en")   = WSubField(rw_exp, "wmode", bt, FEMALE)
-            repl(s"${mem.name}.$w.clk")  = EmptyExpression
-            repl(s"${mem.name}.$w.addr") = EmptyExpression
-            repl(s"${mem.name}.$w.data") = WSubField(rw_exp, "wdata", mem.dataType, FEMALE)
-            repl(s"${mem.name}.$w.mask") = WSubField(rw_exp, "wmask", ut, FEMALE)
-            stmts += Connect(NoInfo, WSubField(rw_exp, "clk", ClockType, FEMALE),
-              WRef("clk", ClockType, NodeKind, MALE))
-            stmts += Connect(NoInfo, WSubField(rw_exp, "en", bt, FEMALE),
-              DoPrim(Or, List(connects(s"${mem.name}.$r.en"), connects(s"${mem.name}.$w.en")), Nil, bt))
-            stmts += Connect(NoInfo, WSubField(rw_exp, "addr", ut, FEMALE),
-              Mux(connects(s"${mem.name}.$w.en"), connects(s"${mem.name}.$w.addr"), 
-                  connects(s"${mem.name}.$r.addr"), ut))
-          }
-        }
-        if (readwriters.isEmpty) mem else DefMemory(mem.info,
-          mem.name, mem.dataType, mem.depth, mem.writeLatency, mem.readLatency,
-          mem.readers filterNot readers, mem.writers filterNot writers,
-          mem.readwriters ++ readwriters)
+  def replaceStmt(repl: Netlist)(s: Statement): Statement =
```

Comment:
I know I basically took this from you for my pass. Maybe this can be made into a pass utility? Same for replaceExp -- although I don't understand why you need to explicitly match to WSubField...